### ISSUE TITLE
Consistently hitting heap limits during optimize-vscode

### ISSUE THREAD

**Issue by Tyriar** [Microsoft/vscode#7108]
- VSCode Version: master
- OS Version: Ubuntu 16.04

The build server doesn't seem to be experiencing the same problem, I tried restarting and re-cloning vscode and it was the same.

```
$ gulp vscode-linux-x64-build-deb
...
[15:21:21] Starting 'optimize-vscode'...

<--- Last few GCs --->

   82576 ms: Mark-sweep 1031.1 (1434.8) -> 1030.9 (1434.8) MB, 744.4 / 0 ms [allocation failure] [GC in old space requested].
   83317 ms: Mark-sweep 1030.9 (1434.8) -> 1030.9 (1434.8) MB, 741.0 / 0 ms [allocation failure] [GC in old space requested].
   84076 ms: Mark-sweep 1030.9 (1434.8) -> 1030.9 (1434.8) MB, 758.3 / 0 ms [last resort gc].
   84831 ms: Mark-sweep 1030.9 (1434.8) -> 1030.9 (1434.8) MB, 755.7 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0xcab8d9c9e59 <JS Object>
    1: _parseMappings(aka SourceMapConsumer_parseMappings) [/home/daimms/dev/Microsoft/vscode/node_modules/concat-with-sourcemaps/node_modules/source-map/lib/source-map-consumer.js:~423] [pc=0x26dfc44e34d1] (this=0x90dba049d09 <a BasicSourceMapConsumer with map 0x26b0e916b2a9>,aStr=0x90dba04a951 <Very long string[1943]>,aSourceRoot=0x90dba049989 <String[44]: file:///home/daimms/dev/Microsoft/vsc...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
Aborted (core dumped)
```

**Comment by jrieken** [comment]
@Tyriar Do you run node in 32 or 64bit? The former has a heap limit of around 1GB, after that you get into GC hell

**Comment by Tyriar** [comment]
@jrieken not sure, sounds reasonable though. I'll check it out when I'm in the office.

**Comment by Tyriar** [comment]
Definitely 64 bit and still hitting this on my laptop after syncing:

```
~
❯ node
> require('os').arch()
'x64'
> 
```

**Comment by jrieken** [comment]
I am not they guy for optimise. Maybe @joaomoreno or @alexandrudima

**Comment by joaomoreno** [comment]
Yeah I've seen this in my Linux desktop, although never in our build machines...

**Comment by joaomoreno** [comment]
Are you on node 5.10?

**Comment by Tyriar** [comment]
@joaomoreno potentially, I'll have to check. Could ./scripts/npm.sh, code.sh and/or gulp check the node version and warn on this if so?

**Comment by alexandrudima** [comment]
If will soon delete editor/standalone-languages (14.7KLOC). Maybe that will help

**Comment by joaomoreno** [comment]
@Tyriar still an issue?

**Comment by Tyriar** [comment]
Nope, I may have been on an old version of node, I'm on 6.2.2 now and it works.

**Comment by mattdsteele** [comment]
Hi folks, I just ran into the same issue. VSCode version: 1.3.1
Platform: Windows 7, x64:

```
C:\Users\matt\code\github\vscode>node -v
v6.3.1

C:\Users\matt\code\github\vscode>node
> require('os').arch()
'x64'
```

While trying to build via `gulp vscode-win32`:

```
[11:05:15] Starting 'optimize-vscode'...

<--- Last few GCs --->

  144155 ms: Mark-sweep 1043.6 (1434.8) -> 1043.6 (1434.8) MB, 1260.3 / 0 ms [allocation failure] [GC in old space reque
sted].
  145415 ms: Mark-sweep 1043.6 (1434.8) -> 1043.5 (1434.8) MB, 1259.8 / 0 ms [allocation failure] [GC in old space reque
sted].
  146690 ms: Mark-sweep 1043.5 (1434.8) -> 1043.5 (1434.8) MB, 1275.1 / 0 ms [last resort gc].
  147930 ms: Mark-sweep 1043.5 (1434.8) -> 1043.5 (1434.8) MB, 1239.8 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0000033BE92C9E51 <JS Object>
    1: _parseMappings(aka SourceMapConsumer_parseMappings) [C:\Users\matt\code\github\vscode\node_modules\concat-with
-sourcemaps\node_modules\source-map\lib\source-map-consumer.js:483] [pc=0000031EBD45EEA4] (this=00000008B10C35B1 <a Basi
cSourceMapConsumer with map 000003658B9F7DE1>,aStr=00000008B10CE2D9 <Very long string[10520]>,aSourceRoot=00000008B10C32
49 <String[47]: file:///C:/Users/IG...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
```

**Comment by Tyriar** [comment]
@mattdsteele I think we switched the version to 1.4.0 almost a month ago, does it happen on `master`?

**Comment by mattdsteele** [comment]
Just rebuilt, and it didn't occur for me on `master`. Looks like it might have been resolved.

**Comment by Tyriar** [comment]
This has been happening again for me on master/Windows 10/node 6.9.2.

**Comment by Tyriar** [comment]
V8 has a strict heap limit of 1.7gb, it can be increased like so:

`node --max-old-space-size=4096 ./node_modules/gulp/bin/gulp.js watch`
`gulp watch --max-old-space-size=4096` should also work

**Comment by Tyriar** [comment]
We should consider either adding this to the `npm run watch` script and/or adding it to the build wiki page

**Comment by joaomoreno** [comment]
Yeah. @isidorn has this too. I hit it a couple of times too.

We should add it and inform everyone to use `npm run watch` instead of `gulp watch`.

**Comment by headmelted** [comment]
This is still happening for me on the latest merge into my ARM builds (although it's happening for Intel 32/64 as well).

I'm trying to tune max_old_space_size and max_new_space_size ( <-- although I'm fairly sure max_new_space_size has nothing to do with the problem ).

Did Tyriar's commit (3e418e1) resolve this for you guys?

**Comment by joaomoreno** [comment]
@headmelted Yes, make sure you use `npm run watch`.

**Comment by headmelted** [comment]
Testing that out now.

Should the package.json also be changed to use "npm run watch"?

It's strange to me that your master builds are still running with the package.json calling "gulp watch" and mine are not, and it makes me slightly suspicious I have something else out of whack.

In any case, I'll keep digging at this side.  Thanks!

**Comment by joaomoreno** [comment]
> Should the package.json also be changed to use "npm run watch"?

What do you mean? `npm run watch` will run what's defined in package.json: `gulp watch --max_old_space_size=4096`.

I'll change our builds to run `npm run watch`. The build fails occasionally, not every time.

**Comment by headmelted** [comment]
I've now got to the bottom of my issue. Tyriar's above fix solved my immediate problem, but it was still falling over, which after checking my build logs turns out to have been due the Linux packaging scripts also failing for the same reason (*-build-deb).

In my case, I needed to increase the heap limit for building the packages, too, i.e.:

**gulp "vscode-linux-armv7l-build-deb" --silent --max_old_space_size=4096;**

It may be worth adding this to your travis script for the subsequent node child scripts, or better yet, setting the increased heap size globally for the duration of the run.
